be/src/exec/spill/spiller.hpp

```diff
@@ -63,72 +125,241 @@ Status Spiller::flush(RuntimeState* state, TaskExecutor&& executor, MemGuard&& g
-
-    ChunkPtr chunk;
-    // @TODO(silverbullet233): reuse ctx
-    SerdeContext ctx;
-    ASSIGN_OR_RETURN(chunk, _input_stream->get_next(ctx));
-    COUNTER_UPDATE(_metrics.restore_rows, chunk->num_rows());
-    _restore_read_rows += chunk->num_rows();
+StatusOr<ChunkPtr> SpillerReader::restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
+    SCOPED_TIMER(_spiller->metrics().restore_timer);
+    ASSIGN_OR_RETURN(auto chunk, _current_stream->get_next(_spill_read_ctx));
     RETURN_IF_ERROR(trigger_restore(state, std::forward<TaskExecutor>(executor), std::forward<MemGuard>(guard)));
+    _read_rows += chunk->num_rows();
     return chunk;
 }
 
 template <class TaskExecutor, class MemGuard>
-Status Spiller::trigger_restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
-    RETURN_IF_ERROR(_get_spilled_task_status());
-    RETURN_IF_ERROR(_acquire_input_stream(state));
-
-    DCHECK(_input_stream->enable_prefetch());
+Status SpillerReader::trigger_restore(RuntimeState* state, TaskExecutor&& executor, MemGuard&& guard) {
+    if (_current_stream == nullptr) {
+        return Status::OK();
+    }
+    DCHECK(_current_stream->enable_prefetch());
     // if all is well and input stream enable prefetch and not eof
-    if (!_input_stream->eof()) {
+    if (!_current_stream->eof()) {
         auto restore_task = [this, state, guard]() {
+            RETURN_IF(!guard.scoped_begin(), Status::OK());
             _running_restore_tasks++;
-            guard.scoped_begin();
-            // @TODO(silverbullet233): reuse ctx
             SerdeContext ctx;
-            auto res = _input_stream->prefetch(ctx);
+            auto res = _current_stream->prefetch(ctx);
+
+            if (!res.is_end_of_file() && !res.ok()) {
+                _spiller->update_spilled_task_status(std::move(res));
+            }
 
-            _update_spilled_task_status(res.is_end_of_file() ? Status::OK() : res);
             guard.scoped_end();
             _running_restore_tasks--;
             if (!res.ok()) {
                 _finished_restore_tasks++;
-                return;
             }
+            return Status::OK();
         };
         RETURN_IF_ERROR(executor.submit(std::move(restore_task)));
     }
     return Status::OK();
 }
-} // namespace spill
-} // namespace starrocks
\ No newline at end of file
+
+template <class TaskExecutor, class MemGuard>
+Status PartitionedSpillerWriter::spill(RuntimeState* state, const ChunkPtr& chunk, TaskExecutor&& executor,
+                                       MemGuard&& guard) {
+    DCHECK(!chunk->is_empty());
+    DCHECK(!is_full());
+
+    // the last column was hash column
+    auto hash_column = chunk->columns().back();
+
+    {
+        SCOPED_TIMER(_spiller->metrics().shuffle_timer);
+        std::vector<uint32_t> shuffle_result;
+        shuffle(shuffle_result, down_cast<SpillHashColumn*>(hash_column.get()));
+        process_partition_data(chunk, shuffle_result,
```

Comment:
it will be used in future (in SpillableHashJoinProbeOperator)
https://github.com/StarRocks/starrocks/pull/19122/files#diff-e35d0a5ac92b2f29f5cedec3b6e4d0475b7a640fe7659f85e6de9723c5b71792R203-R217